### PR TITLE
romLoader and fileIO created, test running

### PR DESCRIPTION
**Pull request by MarjorieLV**

### CODE REVIEW

File: includes/romloader.hpp [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+#define _ROMLOADER_
+
+#include <string>
+#include "fileio.hpp"
+
+class RomLoader
+{
+public:
+    RomLoader(IFileIO& fio);
+    bool load(std::string romName);
```
Comment (shulrak) [inline includes/romloader.hpp]: const &

**Review by shulrak — DISMISSED** [review]

File: includes/ifileio.hpp [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+#define _IFILEIO_
+
+#include <string>
+#include <vector>
+#include <cstdint>
+
+class IFileIO
+{
+public:
+    virtual int openFile(std::string fileName) = 0;
```
Comment (Sgregory42) [inline includes/ifileio.hpp]: const &

File: includes/fileio.hpp [hunk 3]
```diff
@@ -0,0 +1,13 @@
+#ifndef _FILEIO_
+#define _FILEIO_
+
+#include "ifileio.hpp"
+class FileIO : public IFileIO
+{
+public:
+
+    int openFile(std::string fileName);
```
Comment (Sgregory42) [inline includes/fileio.hpp]: const &

File: includes/ifileio.hpp [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+    virtual int readFile(uint8_t *buff, int fd) = 0;
```
Comment (Sgregory42) [inline includes/ifileio.hpp]: Add some doxygen comment. eg :
int sum(int a, int b);
```
/**
       * A pure function doing a sum of two int variable.
       * @param a an integer argument.
       * @param s an integer arguement
       * @return the sum of \p a and \p s
*/
```
so your readFile would look like this : 
```
int readFile(uint8_t *buff, int fd) = 0;
/**
       * pure Virtual function that reads a file descriptor.
       * @param buff pointer to the output buffer. it will contains what we will read.
       * @param fd the filedescriptor we will read from.
       * @return an int, the number of character read. return zero if there no more character to read.
*/
```

Btw how come you are not specifying how many character you are reading ? 

Maybe don't use a pointer but pass a reference to a vector ? (c++ style)

Comment (Sgregory42) [reply includes/ifileio.hpp]: btw do this for most of your classes. usually it's mostly public member function (since they will be used by "others")  :)

File: includes/iromloader.hpp [hunk 5]
```diff
@@ -0,0 +1,13 @@
+#ifndef _IROMLOADER_
+#define _IROMLOADER_
+
+#include <string>
+#include "fileio.hpp"
+
+class IRomLoader
+{
+public:
+    virtual bool load(std::string romName) = 0;
```
Comment (Sgregory42) [inline includes/iromloader.hpp]: const &

File: src/fileio.cpp [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+#include <iostream>
+#include <vector>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include "fileio.hpp"
+
+int FileIO::openFile(std::string fileName)
+{
+    int fd = open("./roms/Pokemon_Bleue.gb", O_RDONLY);
```
Comment (Sgregory42) [inline src/fileio.cpp]: filename not used ?

File: src/fileio.cpp [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+        std::string error = strerror(errno);
+        std::cout << "error open : " << error << std::endl;
+    }
+    return fd;
+}
+
+
+int FileIO::readFile(uint8_t *buff, int fd)
+{
+    int ret = read(fd, buff, 512);
```
Comment (Sgregory42) [inline src/fileio.cpp]: return read(); is better here. the int ret is  just noise.

File: src/fileio.cpp [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+#include <fcntl.h>
+#include <unistd.h>
+#include "fileio.hpp"
+
+int FileIO::openFile(std::string fileName)
+{
+    int fd = open("./roms/Pokemon_Bleue.gb", O_RDONLY);
+    if (fd < 0) {
+        std::string error = strerror(errno);
+        std::cout << "error open : " << error << std::endl;
```
Comment (Sgregory42) [inline src/fileio.cpp]: Add //TODO comment to easily find this error so that when you'll make a better error handler you won't miss it.

File: src/main.cpp [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    //   int fdOutput = open("name", O_WRONLY | O_CREAT, 0666);
+    //   if (fdOutput >= 0) {
+    //     std::string name(memory.begin() + nameBegin, memory.begin() + nameEnd);
+    //     std::cout << name << std::endl;
+    //   }
+    // }
+    // else {
+    //   std::string error = strerror(errno);
+    //   std::cout << "error open : " << error << std::endl;
+    // }
```
Comment (Sgregory42) [inline src/main.cpp]: remove dead code

File: src/romloader.cpp [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+RomLoader::RomLoader(IFileIO& fio)
+    :fileIO(fio){}
+
+bool RomLoader::load(std::string romName)
+{
+    int fd = fileIO.openFile(romName);
+    if (fd > 0) {
+        uint8_t buff[512];
+        while (int readBit = fileIO.readFile(buff, fd)) {
+            for (int i = 0; i < readBit; i++) {
```
Comment (Sgregory42) [inline src/romloader.cpp]: you could something better instead of the for + push_back 
`std::copy(buff, buff + readBit, std::back_inserter(_data));`
there is also
`_data.insert(_data.end(), buff, buff + readBit)` 
I think the latest one is more efficiency (since he knows the size it will _data.reserve(). But anyway it doesn't really matter. the ROM file are small.

**Review by Sgregory42 — CHANGES_REQUESTED** [review]
Where are the tests ?

File: includes/cpu.hpp [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+
+#include <map>
+
+class Cpu
+{
+public:
+
+private:
+    //internal state values
+    std::map<char, uint8_t>
```
Comment (Sgregory42) [inline includes/cpu.hpp]: this file shouldn't be part of this commit

File: test/romloader.t.cpp [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+    MOCK_METHOD1(openFile, int(std::string const &));
+    MOCK_METHOD2(readFile, int(uint8_t*, int));
+    MOCK_METHOD1(closeFile, void(int));
+};
+
+class RomLoaderTest : public ::testing::Test
+{
+public:
+    MockFileIo fileIO;
+    std::string fileName = "../roms/Pokemon_Bleue.gb";
```
Comment (Sgregory42) [inline test/romloader.t.cpp]: use a dummy name instead of a real one

File: test/romloader.t.cpp [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+    EXPECT_CALL(fileIO, readFile(_, fd))
+        .WillOnce(Return(0));
+
+    EXPECT_CALL(fileIO, closeFile(fd));
+
+    RomLoader romLoader(fileIO);
+    bool isLoaded = romLoader.load(fileName);
+
+    EXPECT_TRUE(romLoader.getData().empty());
+    EXPECT_TRUE(isLoaded);
```
Comment (Sgregory42) [inline test/romloader.t.cpp]: Does it make sense to return true if the file is empty ?

**Review by Sgregory42 — CHANGES_REQUESTED** [review]
Minor changes

**Review by Sgregory42 — APPROVED** [review]
lgm

**Review by Sgregory42 — APPROVED** [review]
look good to me
